Userland/DevTools/UserspaceEmulator/SoftFPU.cpp

```diff
@@ -0,0 +1,1744 @@
+    m_fpu_error_summary = 1;
+
+    // FIXME: set traceback
+    // For that we need to get the currently executing instruction and
+    // the previous eip
+
+    // FIXME: Call FPU Exception handler
+    reportln("Trying to call Exception handler from {}", fpu_exception_string(ex));
+    fpu_dump_env();
+    m_emulator.dump_backtrace();
+    TODO();
+}
+
+template<Arithmetic T>
+ALWAYS_INLINE T SoftFPU::fpu_round(long double value) const
+{
+    // FIXME: may need to set indefinite values manually
+    switch (fpu_get_round_mode()) {
+    case RoundingMode::NEAREST:
+        return static_cast<T>(roundl(value));
+    case RoundingMode::DOWN:
+        return static_cast<T>(floorl(value));
+    case RoundingMode::UP:
+        return static_cast<T>(ceill(value));
+    case RoundingMode::TRUNK:
+        return static_cast<T>(truncl(value));
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+template<Arithmetic T>
+ALWAYS_INLINE T SoftFPU::fpu_round_checked(long double value)
+{
+    T result = fpu_round<T>(value);
+    if (auto rnd = value - result) {
+        if (rnd > 0)
+            set_c1(1);
+        else
+            set_c1(0);
+        fpu_set_exception(FPU_Exception::Precision);
+    }
+    return result;
+}
+
+template<FloatingPoint T>
+ALWAYS_INLINE T SoftFPU::fpu_convert(long double value) const
+{
+    // FIXME: actually round the right way
+    return static_cast<T>(value);
+}
+template<FloatingPoint T>
+ALWAYS_INLINE T SoftFPU::fpu_convert_checked(long double value)
+{
+    T result = fpu_convert<T>(value);
+    if (auto rnd = value - result) {
+        if (rnd > 0)
+            set_c1(1);
+        else
+            set_c1(0);
+        fpu_set_exception(FPU_Exception::Precision);
+    }
+    return result;
+}
+
+// Instructions
+
+// DATA TRANSFER
+void SoftFPU::FLD_RM32(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_push(fpu_get(insn.modrm().register_index()));
+    } else {
+        auto new_f32 = insn.modrm().read32(m_cpu, insn);
+        // FIXME: Respect shadow values
+        fpu_push(bit_cast<float>(new_f32.value()));
+    }
+}
+void SoftFPU::FLD_RM64(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto new_f64 = insn.modrm().read64(m_cpu, insn);
+    // FIXME: Respect shadow values
+    fpu_push(bit_cast<double>(new_f64.value()));
+}
+void SoftFPU::FLD_RM80(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+
+    // long doubles can be up to 128 bits wide in memory for reasons (alignment) and only uses 80 bits of precision
+    // GCC uses 12 bytes in 32 bit and 16 bytes in 64 bit mode
+    // so in the 32 bit case we read a bit to much, but that shouldn't be an issue.
+    // FIXME: Respect shadow values
+    u128 new_f80 = insn.modrm().read128(m_cpu, insn).value();
+
+    fpu_push(*(long double*)new_f80.bytes().data());
+}
+
+void SoftFPU::FST_RM32(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
```

Comment:
Why not? Does not make intuitive sense to me.

Reply:
You cant store a floating point value directly into a register, and you really should not